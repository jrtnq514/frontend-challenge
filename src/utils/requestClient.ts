import ky from "ky";

const API_BASE_URL = process.env.API_BASE_URL || "http://localhost:3001/api";

export default ky.create({
  headers: {},
  prefixUrl: API_BASE_URL,
});
