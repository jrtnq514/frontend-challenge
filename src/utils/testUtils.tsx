import { InitialEntry, RouterState } from "@remix-run/router";
import {
  createMemoryRouter,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import {
  SignUpState,
  SignUpStateProvider,
} from "src/contexts/signUpStateContext";
import { render } from "@testing-library/react";

interface RenderViewProps {
  routes: RouteObject[];
  routerOptions?:
    | {
        basename?: string | undefined;
        hydrationData?:
          | Partial<Pick<RouterState, "errors" | "loaderData" | "actionData">>
          | undefined;
        initialEntries?: InitialEntry[] | undefined;
        initialIndex?: number | undefined;
      }
    | undefined;
  signUpState?: SignUpState;
}

export const renderView = ({
  routes,
  routerOptions = {},
  signUpState,
}: RenderViewProps) => {
  const router = createMemoryRouter(routes, routerOptions);

  return {
    ...render(
      <SignUpStateProvider initState={signUpState}>
        <RouterProvider router={router} />
      </SignUpStateProvider>
    ),
    router,
  };
};
