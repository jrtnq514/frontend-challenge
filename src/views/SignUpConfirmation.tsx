import React, { FormEvent, useEffect, useState } from "react";
import {
  Box,
  Spinner,
  VStack,
  Heading,
  Text,
  List,
  ListItem,
  HStack,
  Button,
} from "@chakra-ui/react";
import { useNavigate, useNavigation } from "react-router-dom";
import { useSignUpState } from "src/contexts/signUpStateContext";
import request from "src/utils/requestClient";

const SignUpConfirmation: React.FC = () => {
  const navigate = useNavigate();
  const { state: navState } = useNavigation();
  const { signUpState } = useSignUpState();
  const [showPassword, setShowPassword] = useState(false);
  const [isSubmitting, setIsSubmitting] = useState(false);

  useEffect(() => {
    if (!signUpState.firstName || !signUpState.email || !signUpState.password) {
      navigate("/");
    }

    if (!signUpState.favoriteColor || !signUpState.agreeTermsAndConditions) {
      navigate("/more-info");
    }
  }, []);

  const submitSignUp = async (event: FormEvent) => {
    event.preventDefault();
    setIsSubmitting(true);
    try {
      await request.post("submit", {
        json: {
          name: signUpState.firstName,
          email: signUpState.email,
          password: signUpState.password,
          color: signUpState.favoriteColor,
          terms: signUpState.agreeTermsAndConditions,
        },
      });
      navigate("/success");
    } catch (error) {
      console.log(error);
      navigate("/error");
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <Box>
      {navState === "loading" ? (
        <Spinner size="xl" />
      ) : (
        <Box
          as="form"
          data-testid="confirmation-form"
          width="100%"
          onSubmit={submitSignUp}
        >
          <VStack spacing={6}>
            <Heading as="h1">Confirmation</Heading>
            <Text>Make sure everything looks ok.</Text>
            <List fontSize="xl" spacing={4}>
              <ListItem>
                <HStack wrap="wrap">
                  <Box fontWeight="light">First Name:</Box>
                  <Box data-testid="first-name-value">
                    {signUpState.firstName}
                  </Box>
                </HStack>
              </ListItem>
              <ListItem>
                <HStack wrap="wrap">
                  <Box fontWeight="light">Email:</Box>
                  <Box data-testid="email-value">{signUpState.email}</Box>
                </HStack>
              </ListItem>
              <ListItem>
                <HStack wrap="wrap">
                  <Box fontWeight="light">Password:</Box>
                  <Box data-testid="password-value">
                    {showPassword
                      ? signUpState.password
                      : signUpState.password?.split("").map((x) => `â€¢`)}
                  </Box>
                  <Button
                    data-testid="show-password-btn"
                    size="sm"
                    variant="outline"
                    onClick={(event) => setShowPassword(!showPassword)}
                  >
                    {showPassword ? "Hide" : "Show"}
                  </Button>
                </HStack>
              </ListItem>
              <ListItem>
                <HStack wrap="wrap">
                  <Box fontWeight="light">Favorite Color:</Box>
                  <Box data-testid="color-value">
                    {signUpState.favoriteColor}
                  </Box>
                </HStack>
              </ListItem>
              <ListItem>
                <HStack wrap="wrap">
                  <Box fontWeight="light">Accepted Terms and Conditions: </Box>
                  <Box data-testid="terms-value">
                    {signUpState.agreeTermsAndConditions
                      ? "Accepted"
                      : "Declined"}
                  </Box>
                </HStack>
              </ListItem>
            </List>
            <HStack justifyContent="space-between" width="100%">
              <Button
                type="button"
                variant="outline"
                colorScheme="green"
                onClick={(event) => navigate("/more-info")}
              >
                Back
              </Button>
              <Button
                data-testid="submit-btn"
                type="submit"
                variant="solid"
                colorScheme="green"
                isDisabled={isSubmitting}
              >
                {isSubmitting ? (
                  <Spinner data-testid="submit-btn-spinner" />
                ) : (
                  "Submit"
                )}
              </Button>
            </HStack>
          </VStack>
        </Box>
      )}
    </Box>
  );
};

export default SignUpConfirmation;
