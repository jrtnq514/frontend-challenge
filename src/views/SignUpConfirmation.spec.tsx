import SignUpConfirmation from "./SignUpConfirmation";
import { waitFor, screen, fireEvent } from "@testing-library/react";
import { renderView } from "src/utils/testUtils";

describe("SignUpConfirmation", () => {
  const mockColors = ["Blue", "Black", "Red"];

  it("should render sign up state values", async () => {
    const { container } = renderView({
      routes: [
        {
          path: "/confirmation",
          element: <SignUpConfirmation />,
        },
      ],
      routerOptions: { initialEntries: ["/confirmation"] },
      signUpState: {
        firstName: "test",
        email: "test",
        password: "test",
        favoriteColor: "test",
        agreeTermsAndConditions: true,
      },
    });

    await waitFor(() =>
      Promise.all([
        screen.findByTestId("first-name-value"),
        screen.findByTestId("email-value"),
        screen.findByTestId("password-value"),
        screen.findByTestId("color-value"),
        screen.findByTestId("terms-value"),
      ])
    );

    const renderedValues = await Promise.all([
      screen.findByTestId("first-name-value"),
      screen.findByTestId("email-value"),
      screen.findByTestId("password-value"),
      screen.findByTestId("color-value"),
      screen.findByTestId("terms-value"),
    ]);

    expect(renderedValues.map((x) => x.textContent)).toEqual([
      "test",
      "test",
      "••••",
      "test",
      "Accepted",
    ]);
    expect(container).toMatchSnapshot();
  });

  it("should show a loading spinner on submit btn click", async () => {
    renderView({
      routes: [
        {
          path: "/confirmation",
          element: <SignUpConfirmation />,
        },
        {
          path: "/success",
          element: <div>Success</div>,
        },
      ],
      routerOptions: { initialEntries: ["/confirmation"] },
      signUpState: {
        firstName: "test",
        email: "test",
        password: "test",
        favoriteColor: "test",
        agreeTermsAndConditions: true,
      },
    });

    await waitFor(() => screen.findByTestId("submit-btn"));

    const submitBtn = await screen.findByTestId("submit-btn");
    fireEvent.click(submitBtn);

    expect(await screen.findByTestId("submit-btn-spinner")).not.toBe(null);
  });
});
