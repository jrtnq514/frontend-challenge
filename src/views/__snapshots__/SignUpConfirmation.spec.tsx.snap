// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpConfirmation should render sign up state values 1`] = `
<div>
  <div
    class="css-0"
  >
    <form
      class="css-8atqhb"
      data-testid="confirmation-form"
    >
      <div
        class="chakra-stack css-pfcv3f"
      >
        <h1
          class="chakra-heading css-0"
        >
          Confirmation
        </h1>
        <p
          class="chakra-text css-0"
        >
          Make sure everything looks ok.
        </p>
        <ul
          class="css-16au9ar"
          role="list"
        >
          <li
            class="css-0"
          >
            <div
              class="chakra-stack css-uxb30m"
            >
              <div
                class="css-13xnkip"
              >
                First Name:
              </div>
              <div
                class="css-0"
                data-testid="first-name-value"
              >
                test
              </div>
            </div>
          </li>
          <li
            class="css-0"
          >
            <div
              class="chakra-stack css-uxb30m"
            >
              <div
                class="css-13xnkip"
              >
                Email:
              </div>
              <div
                class="css-0"
                data-testid="email-value"
              >
                test
              </div>
            </div>
          </li>
          <li
            class="css-0"
          >
            <div
              class="chakra-stack css-uxb30m"
            >
              <div
                class="css-13xnkip"
              >
                Password:
              </div>
              <div
                class="css-0"
                data-testid="password-value"
              >
                •
                •
                •
                •
              </div>
              <button
                class="chakra-button css-4xx2wk"
                data-testid="show-password-btn"
                type="button"
              >
                Show
              </button>
            </div>
          </li>
          <li
            class="css-0"
          >
            <div
              class="chakra-stack css-uxb30m"
            >
              <div
                class="css-13xnkip"
              >
                Favorite Color:
              </div>
              <div
                class="css-0"
                data-testid="color-value"
              >
                test
              </div>
            </div>
          </li>
          <li
            class="css-0"
          >
            <div
              class="chakra-stack css-uxb30m"
            >
              <div
                class="css-13xnkip"
              >
                Accepted Terms and Conditions: 
              </div>
              <div
                class="css-0"
                data-testid="terms-value"
              >
                Accepted
              </div>
            </div>
          </li>
        </ul>
        <div
          class="chakra-stack css-1cqqob7"
        >
          <button
            class="chakra-button css-4xx2wk"
            type="button"
          >
            Back
          </button>
          <button
            class="chakra-button css-4xx2wk"
            data-testid="submit-btn"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
